### PR TITLE
Credit note

### PR DESCRIPTION
**Pull request by noplisu**
## Pull Request Summary
We want to be able to correct invoices send into the KSeF API. To correct the invoices we need to prepare a credit note.

There are a few differences between a regular VAT invoice and a polish credit note.
- The first is the InvoiceType that changes from VAT to KOR.
- The second is a CorrectionType it is used to define when the correction should take legal effect, with values 1-3 meaning:
1 - Changes take effect in the date of the invoice. This is mostly used if you made a mistake in the original invoice.
2 - Changes take effect in the date of the correction. This is used for example when giving a discount or returning items.
3 - Changes take effect in a different date than any of the invoices or the date is different for different elements.
- There is also a optional field CorrectionReason where you can give a reason for the correction.
- The last element is data of the Invoice you are correcting. You need to give the SequentialNumber and IssueDate of the original invoice. If the invoice has a KSeF number you need to provide it and set the KsefNumberPresent flag to 1. If the original invoice didn't have a KSeF number you need to set the NoKsefNumberPresent flag to 1.

## Dependencies
- This PR requires changes in GOBL: https://github.com/invopop/gobl/pull/254

## Sources
- [FA2 structure](https://www.podatki.gov.pl/media/9466/dokumentacja-struktury-logicznej-e-faktury-fa-2.pdf)
- [VAT act on correction invoices](https://sip.lex.pl/akty-prawne/dzu-dziennik-ustaw/podatek-od-towarow-i-uslug-17086198/art-106-j)

### CODE REVIEW

**Comment by samlown** [comment]
The descriptions and links in this PR are great! Help a lot to find source data. However, its not clear to me how "Corrective" invoices work in Poland. Its not totally clear to me if the corrective document *replaces* the previous document, or *amends*.

File: lines.go [hunk 1]
```diff
@@ -27,14 +27,24 @@ func newLine(line *bill.Line) *Line {
...
 }
 
+func unitDiscount(line *bill.Line) string {
+	discount := line.Total.Subtract(line.Sum).Divide(line.Quantity) // not sure if there should be some rescale or matchPrecision
+
+	if discount.IsZero() {
+		return ""
+	}
+
+	return discount.String()
```
Comment (torrocus) [inline lines.go:45]: What type of discount is it? I ask because since we perform mathematical operations (subtraction and division), these calculations are probably sensitive to precision.

If I understand correctly, the `discount` will be 0 if `line.Total` is equal to `line.Sum`. We can check this at the very beginning of the function.

```suggestion
	if line.Total == line.Sum {
		return ""
	}

	discount := line.Total.Subtract(line.Sum).Divide(line.Quantity) // not sure if there should be some rescale or matchPrecision

	return discount.String()
```
What do you think?

Comment (noplisu) [reply lines.go]: Yes, this is a better approach. 👍

File: lines.go [hunk 2]
```diff
@@ -27,14 +27,24 @@ func newLine(line *bill.Line) *Line {
...
 
 	return Line
 }
 
+func unitDiscount(line *bill.Line) string {
+	if line.Total == line.Sum {
+		return ""
+	}
+
+	discount := line.Total.Subtract(line.Sum).Divide(line.Quantity) // not sure if there should be some rescale or matchPrecision
```
Comment (samlown) [inline lines.go:43]: Can you not take the Discount amount here and divide by the quantity instead of using the total? Under normal circumstances these types of calculation can easily cause rounding errors, I'd assume however that KSeF are not recalculating the totals.

File: test/data/corrective-invoice.json [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    "uuid": "d05c7988-e145-11ee-bef0-52f0b4d06139",
+    "dig": {
+      "alg": "sha256",
+      "val": "0630d41061873a3b8bf650cbfd204a57c6157210d8a6f49c4a4ee31338b43e1f"
+    },
+    "draft": true
+  },
+  "doc": {
+    "$schema": "https://gobl.org/draft-0/bill/invoice",
+    "type": "corrective",
```
Comment (samlown) [inline test/data/corrective-invoice.json:13]: I'd expect this to be `credit-note` and the rest of the numbers in the document to not be negative.

**Review by samlown — CHANGES_REQUESTED** [review]
I'd be important to get the tests and linter passing here also!

**Comment by noplisu** [comment]
@samlown Yeah, The tests and linters will pass after I merge the PR in gobl: https://github.com/invopop/gobl/pull/254 and bump the version.

File: lines.go (hunk 2 above)
Comment (noplisu) [reply lines.go]: I can't find one Discount amount and calculating the Discount from the array adds more calculations. I removed this part for the time being as this is a optional field. It can be tackled in a separate PR.

Comment (samlown) [reply lines.go]: This is true, but semantically it is more accurate. I'm not worried about adding up, the division is what dramatically increases the chances of rounding errors. I'd imagine the field is optional because most lines don't have discounts, but if a discount is applied, not defining this field might cause problems for the totals. Is there another discount field that can be used?

**Comment by samlown** [comment]
> @samlown Yeah, The tests and linters will pass after I merge the PR in gobl: [invopop/gobl#254](https://github.com/invopop/gobl/pull/254) and bump the version.

You can build with a specific branch if needed:

```
go get github.com/invopop/gobl@noplisu:gl-regime-pl-add-correction-invoice-fields
```

File: lines.go (hunk 2 above)
Comment (noplisu) [reply lines.go]: I checked the fields and this is the only field that has to do with discounts. I added a new implementation for the field that is calculated from the discounts.

File: go.mod [hunk 4]
```diff
@@ -36,4 +37,4 @@ require (
 	golang.org/x/crypto v0.17.0 // indirect
 )
 
-// replace github.com/invopop/gobl => ../gobl
+// replace github.com/invopop/gobl => github.com/noplisu/gobl v0.0.0-20230327142000-57a48df89539
```
Comment (torrocus) [inline go.mod:40]: [idea] Once the new changes are implemented in invopop/gobl, this change won't be needed, right? Perhaps it makes sense for it to be in a separate commit. Then we can drop it easily.

**Review by torrocus — APPROVED** [review]
